### PR TITLE
Memory safety

### LINKED ISSUE #9: memory safety issue: free(): invalid pointer
Hello,
  Recently a CI run in good_lp returned the following, I think this is a memory safety bug in the coin_cbc interface. I cannot reproduce the issue on my machine, but here is the CI run:

  https://github.com/rust-or/good_lp/pull/2/checks

```
---- src/variables_macro.rs - variables (line 14) stdout ----
Test executable failed (terminated by signal).

stdout:
Dual infeasible - objective value 0
DualInfeasible objective 0 - 0 iterations time 0.002

stderr:
free(): invalid pointer
``` 
[github actions runner logs.zip](https://github.com/KardinalAI/coin_cbc/files/6082169/github.actions.runner.logs.zip)

good_lp uses only safe functions in coin_cbc.

### PR DESCRIPTION
**Pull request by lovasoa**
Fix memory corruption vulnerability

Fixes #9

### CODE REVIEW

File: src/raw.rs [hunk 1]
```diff
@@ -13,6 +13,22 @@ use std::convert::TryInto;
...
 
+#[cfg(feature = "singlethread-cbc")]
+fn lock<T, F: FnMut() -> T>(mut f: F) -> T {
+    lazy_static::lazy_static! {
+        static ref GLOBAL_MUTEX: std::sync::Mutex<()> = std::sync::Mutex::new(());
+    }
+    let lock = GLOBAL_MUTEX.lock().unwrap();
+    let result = f();
+    drop(lock);
+    result
```
Comment (TeXitoi) [inline src/raw.rs]:
```suggestion
    let _lock = GLOBAL_MUTEX.lock().unwrap();
    f()
```

**Comment by lovasoa** [comment]
@dlaehnemann : You are probably more familiar with this issue than I am. I added a lock only around the solve method, which seems to work (I added a test for it). Is that safe, or should the lock be wider ?

File: Cargo.toml [hunk 2]
```diff
@@ -14,5 +14,13 @@ readme = "README.md"
 [workspace]
 members = ["coin_cbc_sys"]
 
+[features]
+default = ["singlethread-cbc"]
+# The `singlethread-cbc` feature MUST be activated if this crate is going to be used with
+# a cbc dynamic library that has not been explicitly compiled with `-DCBC_THRED_SAFE`.
+# For more information, see https://github.com/KardinalAI/coin_cbc/issues/9
```
Comment (TeXitoi) [inline Cargo.toml:21]: Please add this documentation somewhere more visible, as in the README.md and the top level documentation at the head of lib.rs

**Review by TeXitoi — CHANGES_REQUESTED** [review]
Thanks, need a bit of documentation

File: Cargo.toml (hunk 2 above)
Comment (lovasoa) [reply Cargo.toml:21]: @TeXitoi : done in 353c5e3

**Comment by dlaehnemann** [comment]
I am not sure if this is the only place. TBH, I don't understand how exactly the bindings are working. From what I understand, the problem lies in some argument parsing, so this problem might be happening earlier, before a `solve()`. But I'm not sure.

I guess my approach would be to simply run parallel tests in a loop as I suggested in PR #7.

**Comment by lovasoa** [comment]
~~The main problem is not tests. The main problem is that attackers may end up executing arbitrary code on the machine of anyone using this library.~~ 

Edit: I think I initially misunderstood your comment. I thought you were suggesting to run the tests sequentially instead of in parallel, whereas you were actually suggesting the opposite: run the tests in parallel many times to try and expose a potential problem. Right ?

**Comment by lovasoa** [comment]
I would say let's already merge this as a first step, and increase the scope of the lock if we discover more problems. My initial testing seems to indicate that the problem lies only in `Cbc_solve`.

**Comment by lovasoa** [comment]
@dlaehnemann : I did what you suggested, and added more tests to try and exercise more different code paths in parallel. Everything seems to be okay with a lock only around Cbc_solve.

File: src/lib.rs [hunk 3]
```diff
@@ -356,6 +368,36 @@ mod test {
...
+
+    #[test]
+    fn parallel_solves() {
+        // Solve many instances of the knapsack test above, in parallel
+        let knapsacks = (0..50).map(|_| std::thread::spawn(knapsack)).collect::<Vec<_>>();
+        let sos = (0..50).map(|_| std::thread::spawn(with_sos)).collect::<Vec<_>>();
+        for (t1, t2) in knapsacks.into_iter().zip(sos) {
+            t1.join().unwrap();
+            t2.join().unwrap();
+        }
```
Comment (TeXitoi) [inline src/lib.rs]:
```suggestion
        for t in knapsacks.into_iter().chain(sos) {
            t.join().unwrap();
        }
```

File: src/lib.rs [hunk 4]
```diff
@@ -356,6 +368,36 @@ mod test {
...
+    }
+
+    #[test]
+    fn infeasible() {
+        // Solve many instances of the knapsack test above, in parallel
```
Comment (TeXitoi) [inline src/lib.rs]: This comment seems wrong

File: src/raw.rs [hunk 5]
```diff
@@ -13,6 +13,20 @@ use std::convert::TryInto;
 use std::ffi::CStr;
 use std::os::raw::c_int;
 
+#[cfg(feature = "singlethread-cbc")]
+fn lock<T, F: FnMut() -> T>(mut f: F) -> T {
```
Comment (TeXitoi) [inline src/raw.rs]:
```suggestion
fn lock<T, F: FnOnce() -> T>(f: F) -> T {
```

File: src/raw.rs [hunk 6]
```diff
@@ -13,6 +13,20 @@ use std::convert::TryInto;
...
+    lazy_static::lazy_static! {
+        static ref GLOBAL_MUTEX: std::sync::Mutex<()> = std::sync::Mutex::new(());
+    }
+    let _lock = GLOBAL_MUTEX.lock().unwrap();
+    f()
+}
+
+#[cfg(not(feature = "singlethread-cbc"))]
+fn lock<T, F: FnMut() -> T>(mut f: F) -> T {
```
Comment (TeXitoi) [inline src/raw.rs]:
```suggestion
fn lock<T, F: FnOnce() -> T>(f: F) -> T {
```

**Review by TeXitoi — CHANGES_REQUESTED** [review]

**Comment by TeXitoi** [comment]
I might be wrong, but I don't thing there is bugs in calling non-solve method in parallel.

**Comment by dlaehnemann** [comment]
> ~The main problem is not tests. The main problem is that attackers may end up executing arbitrary code on the machine of anyone using this library.~
> 
> Edit: I think I initially misunderstood your comment. I thought you were suggesting to run the tests sequentially instead of in parallel, whereas you were actually suggesting the opposite: run the tests in parallel many times to try and expose a potential problem. Right ?

Yes to the second interpretation. :D
I did this with a bash wrapper, but it obviously makes more sense to do this in the tests, especially as each test does not take long.

And good to hear that it works with the lock. But I'm assuming this essentially does mean that solving is done sequentially for each problem, right? If we wanted to avoid that, the lock would have to be more granularly around the offending code -- but I think we don't get that level of detail from the C API, so I guess this is as good as the solution over here can be.

Otherwise, let's hope for a new release of `Cbc` some time soon. It looks like they have a lot of improvements in the pipeline...

**Comment by TeXitoi** [comment]
This is clearly a workaround, and the proper fix is to have a thread safe cbc lib. There is nothing we can do for finer grain locking.

**Comment by dlaehnemann** [comment]
Just a quick thought: As we will want to revert this whenever `cbc` is fixed permanently, I would squash all the commits in this PR into one merge-commit -- then you can simply `git revert` that single commit later on and have all the code and docs removed at once.

**Comment by lovasoa** [comment]
> As we will want to revert this whenever cbc is fixed permanently,

I don't think we will want to revert that until we can be sure that any user of any program that uses this crate anywhere uses the fixed cbc. Since this library links to cbc dynamically, the version of cbc on the end user's computer will probably be whatever is in their distribution repositories.

**Comment by TeXitoi** [comment]
1. of course i'll squash while merging
2. we have no way of knowing that the lib is updated or not (imagine a debian package that is a few years old), so this will stay for a few years.

**Comment by lovasoa** [comment]
And if this is removed one day, I think only the small `lock` function should be removed, but not all the added tests.

**Review by TeXitoi — APPROVED** [review]

**Comment by TeXitoi** [comment]
v0.1.4 published. Thanks!

**Comment by lovasoa** [comment]
Thanks to both of you, @TeXitoi  and @dlaehnemann  ! 🎉

**Pull request merged**